### PR TITLE
[release/8.0-staging] Set version in ZIP local header to ZIP64 when file offset is >4GB

### PR DESCRIPTION
**Pull request by carlossanlop**
Backport of https://github.com/dotnet/runtime/pull/102053 to release/8.0-staging.
Fixes https://github.com/dotnet/runtime/issues/94899

## Customer Impact

Affecting two partner areas separately.

### dotnet/Open-XML-SDK

Initially reported here: https://github.com/dotnet/wpf/issues/2676

An external customer that develops PowerPoint Add-Ins uses [DocumentFormat.OpenXML](https://github.com/dotnet/wpf/issues/2676) (a package owned by [dotnet/Open-XML-SDK](https://github.com/dotnet/Open-XML-SDK)) to add custom xml parts to the PowerPoint presentation.

When they attempt to reopen the modified PowerPoint file either using [Open XML SDK 2.5 Productivity Tool for Microsoft Office](https://github.com/dotnet/Open-XML-SDK/releases/tag/v2.5) (an asset owned and released also by [dotnet/Open-XML-SDK](https://github.com/dotnet/Open-XML-SDK)) or once again the `DocumentFormat.OpenXML` APIs, they fail stating that "The file contains corrupted data".

But they have no problem opening the file with PowerPoint, or with other zip tools like WinRAR, 7-Zip.

### microsoft/DacFx

This is also affecting our partners at microsoft/DacFx: https://github.com/microsoft/DacFx/issues/363

Azure DevOps uses SqlPackage to export databases into bacpac files and then import them into Azure Sql Server instances.

When they import the bacpac using either Local Sql Server or Azure Sql Server, they have no problems. But when they import it using their Import Database UI or `az sql db import`, there's a failure stating the file contains corrupted data.

All four cases use System.IO.Packaging to generate the bacpac file, but only the failing two cases use System.IO.Packaging again to attempt to import the file.

### Root cause

Both issues have a common root cause: https://github.com/dotnet/runtime/issues/94899

ZipArchive behaves differently in .NET and .NET Framework when creating or reading files larger than 4GB where the algorithm switches to using ZIP64 (described in [this comment](https://github.com/dotnet/runtime/issues/94899#issuecomment-1817756918)). The spec was not clear about that as described in [this comment](https://github.com/dotnet/runtime/issues/94899#issuecomment-1823684588), but other tools do the right thing and can handle the files without issues, so we should match that behavior and ensure there's no discrepancy between .NET and .NET Framework.

The fix consists in making sure that both the central directory header and the local file header store the size offset and when the file is larger than 4GB, set the ZIP64 flag on both, and make sure .NET validates this just as .NET Framework used to do. 

## Testing

New unit tests added:
- In System.IO.Compression, a test that confirms that ZipArchive can open a zip file larger than 4GB that was created with .NET Framework.
- In System.IO.Compression, a test that uses reflection to confirm the correct ZIP64 value is used when creating archives larger than 4GB containing more than one file.
- In System.IO.Packaging, a test similar to the previous one, but using the APIs in this namespace, which wrap System.IO.Compression internally.

## Risk

Low, as the behavior was wrong in .NET when compared to .NET Framework or other ZIP tools.

Thanks to @karakasa for contributing with the original fix.

cc @pwoosam @SeenaAugusty @llali

### CODE REVIEW

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 1]
```diff
@@ -797,7 +797,51 @@ private bool IsOpenable(bool needToUncompress, bool needToLoadIntoMemory, out st
...
+                        CompressionLevel.Fastest => 6,
+                        CompressionLevel.NoCompression => 6,
+                        _ => 0
+                    }
+                    : 0);
+
+            return (BitFlagValues)(((int)generalPurposeBitFlag & ~0x6) | deflateCompressionOptions);
+        }
+
+        private bool IsOffsetTooLarge => _offsetOfLocalHeader > uint.MaxValue;
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:842]: Have you looked to make sure you'll always call this after the value has been set?  Is there an assert you could add to ensure that?

**Review by ericstj — APPROVED** [review]

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:842]: Which value?

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs [hunk 2]
```diff
@@ -797,7 +797,51 @@ private bool IsOpenable(bool needToUncompress, bool needToLoadIntoMemory, out st
             return true;
         }
 
-        private bool SizesTooLarge() => _compressedSize > uint.MaxValue || _uncompressedSize > uint.MaxValue;
+        private bool AreSizesTooLarge => _compressedSize > uint.MaxValue || _uncompressedSize > uint.MaxValue;
+
+        private static CompressionLevel MapCompressionLevel(BitFlagValues generalPurposeBitFlag, CompressionMethodValues compressionMethod)
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: These two methods don't appear to be related to this change.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs [hunk 3]
```diff
@@ -1,48 +1,129 @@
...
+        Random.Shared.NextBytes(buffer);
+    }
 
-            string zipArchivePath = Path.Combine(Path.GetTempPath(), "over4GB.zip");
-            DirectoryInfo tempDir = Directory.CreateDirectory(Path.Combine(Path.GetTempPath(), "over4GB"));
+    [ConditionalTheory(typeof(PlatformDetection), nameof(PlatformDetection.IsSpeedOptimized), nameof(PlatformDetection.Is64BitProcess))] // don't run it on slower runtimes
+    [OuterLoop("It requires 5~6 GB of free disk space and a lot of CPU time for compressed tests")]
+    [InlineData(false)]
+    [InlineData(true)]
+    public static void CheckZIP64VersionIsSet_ForSmallFilesAfterBigFiles(bool isCompressed)
```
Comment (ericstj) [inline src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:56]: Can we also make sure we continue to open the "bad" zips that were produced prior to this fix?  I think we should, but it would be good to capture that in a regression test.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 2 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: thanks for noticing, they're from an unrelated change that was in main (accidentally included when fixing the merge conflict)

Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs]: Fixed

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs (hunk 3 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:56]: Sure. I'm currently writing this test locally. Will submit it as a separate PR in main, and will then backport it, so that I don't block this.

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: No but I can take a look to see if we need to do any small refactoring in this area.

Comment (ericstj) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: I wonder if
```
 private bool IsOffsetTooLarge 
 {
    Debug.Assert(_offsetOfLocalHeader > 0);
    _offsetOfLocalHeader > uint.MaxValue;
}
```
Is enough, or if 0 is a valid value.

File: src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs (hunk 3 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/tests/ZipArchive/zip_LargeFiles.cs:56]: Confirmed in main in this PR: https://github.com/dotnet/runtime/pull/103152

File: src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs (hunk 1 above)
Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: Would you consider this suggestion a blocker for merging this servicing PR, or can I investigate this and fix it in a later change?

Comment (carlossanlop) [reply src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs:802]: I will definitely need more time to investigate this. I quickly tested this suggestion and I am seeing test failures. For now, I think we're ok with checking `IsOffsetTooLarge` even if the underlying field has not been updated. The main scenario that was failing has been confirmed to be fixed.

**Comment by carlossanlop** [comment]
CI failure in osx is an infra issue unrelated to this PR: 
```
/Users/runner/work/1/s/artifacts/toolset/restore.proj : error : Unable to load the service index for source https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-libraries-transport/nuget/v3/index.json.
```
